File: java/tsfile/src/test/java/org/apache/tsfile/read/TsFileTreeReaderRowQueryTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.tsfile.read;

import org.apache.tsfile.enums.TSDataType;
import org.apache.tsfile.exception.write.WriteProcessException;
import org.apache.tsfile.read.query.dataset.ResultSet;
import org.apache.tsfile.read.v4.ITsFileTreeReader;
import org.apache.tsfile.read.v4.TsFileTreeReaderBuilder;
import org.apache.tsfile.write.record.TSRecord;
import org.apache.tsfile.write.schema.MeasurementSchema;
import org.apache.tsfile.write.v4.TsFileTreeWriter;
import org.apache.tsfile.write.v4.TsFileTreeWriterBuilder;

import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

import java.io.File;
import java.io.IOException;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 * Unit tests for {@link ITsFileTreeReader#queryByRow} (tree-model row-range query).
 *
 * <p>Each device has {@value #TOTAL} rows with timestamps 0..TOTAL-1 and values timestamp * 10.
 */
public class TsFileTreeReaderRowQueryTest {

  private static final String FILE_PATH = "test_tree_reader_row_query.tsfile";
  private static final int TOTAL = 50;
  private static final String DEVICE = "device";
  private static final String MEA = "s1";

  private ITsFileTreeReader reader;

  @Before
  public void setUp() throws IOException, WriteProcessException {
    writeTreeFile(FILE_PATH, Collections.singletonList(DEVICE), MEA, TOTAL);
    reader = new TsFileTreeReaderBuilder().file(new File(FILE_PATH)).build();
  }

  @After
  public void tearDown() throws IOException {
    if (reader != null) {
      reader.close();
    }
    new File(FILE_PATH).delete();
  }

  // ① limit=0 → empty result
  @Test
  public void testLimitZeroReturnsEmpty() throws IOException {
    Assert.assertEquals(0, countRows(reader, DEVICE, MEA, 0, 0));
  }

  // ② limit < total → exactly `limit` rows
  @Test
  public void testLimitLessThanTotal() throws IOException {
    Assert.assertEquals(20, countRows(reader, DEVICE, MEA, 0, 20));
  }

  // ③ limit > total → all rows
  @Test
  public void testLimitExceedsTotal() throws IOException {
    Assert.assertEquals(TOTAL, countRows(reader, DEVICE, MEA, 0, 1000));
  }

  // ④ limit=-1 → unlimited, returns all rows
  @Test
  public void testNegativeLimitMeansUnlimited() throws IOException {
    Assert.assertEquals(TOTAL, countRows(reader, DEVICE, MEA, 0, -1));
  }

  // ⑤ offset + limit in the middle
  @Test
  public void testOffsetPlusLimit() throws IOException {
    Assert.assertEquals(15, countRows(reader, DEVICE, MEA, 10, 15));
  }

  // ⑥ offset >= total → empty result
  @Test
  public void testOffsetBeyondTotal() throws IOException {
    Assert.assertEquals(0, countRows(reader, DEVICE, MEA, 1000, 10));
  }

  // ⑦ offset + limit > total → return remaining rows from offset
  @Test
  public void testOffsetPlusLimitExceedsTotal() throws IOException {
    // offset=40, limit=20 → only 10 rows remain
    Assert.assertEquals(10, countRows(reader, DEVICE, MEA, 40, 20));
  }

  // ⑧ data correctness: timestamps and values start from `offset`
  @Test
  public void testDataCorrectness() throws IOException {
    List<String> deviceIds = Collections.singletonList(DEVICE);
    List<String> measurements = Collections.singletonList(MEA);
    ResultSet rs = reader.queryByRow(deviceIds, measurements, 5, 10);
    int count = 0;
    while (rs.next()) {
      long ts = rs.getLong(1); // column 1 = Time
      long val = rs.getLong(2); // column 2 = measurement
      Assert.assertEquals(5 + count, ts);
      Assert.assertEquals((5 + count) * 10L, val);
      count++;
    }
    rs.close();
    Assert.assertEquals(10, count);
  }

  // ⑨ metadata is accessible via the result set
  @Test
  public void testMetadataAccessible() throws IOException {
    List<String> deviceIds = Collections.singletonList(DEVICE);
    List<String> measurements = Collections.singletonList(MEA);
    ResultSet rs = reader.queryByRow(deviceIds, measurements, 0, 5);
    Assert.assertNotNull(rs.getMetadata());
    // Column 1 is always "Time"; column 2 is the full path "device.s1"
    Assert.assertEquals("Time", rs.getMetadata().getColumnName(1));
    Assert.assertEquals(DEVICE + "." + MEA, rs.getMetadata().getColumnName(2));
    rs.close();
  }

  // ⑩ paging consistency: two pages together equal the full result
  @Test
  public void testPaginationConsistency() throws IOException {
    int page1 = countRows(reader, DEVICE, MEA, 0, 25);
    int page2 = countRows(reader, DEVICE, MEA, 25, 25);
    Assert.assertEquals(TOTAL, page1 + page2);
  }

  // ⑪ multiple devices: offset/limit applied to merged result
  @Test
  public void testMultipleDevices() throws IOException, WriteProcessException {
    String filePath = "test_tree_reader_row_query_multi.tsfile";
    writeTreeFile(filePath, Arrays.asList("dev1", "dev2"), MEA, 20);
    try (ITsFileTreeReader r = new TsFileTreeReaderBuilder().file(new File(filePath)).build()) {
      int count = countRows(r, null, MEA, 5, 10);
      Assert.assertEquals(10, count);
    } finally {
      new File(filePath).delete();
    }
  }

  // ─────────────────────────────────────────────────────────────────────────────
  // Helpers
  // ─────────────────────────────────────────────────────────────────────────────

  /**
   * Count rows returned by {@code queryByRow}.
   *
   * @param r reader (already open)
   * @param device device ID; if {@code null}, both "dev1" and "dev2" are queried
   * @param mea measurement name
   * @param offset row offset
   * @param limit row limit
   */
  private int countRows(ITsFileTreeReader r, String device, String mea, int offset, int limit)
      throws IOException {
    List<String> deviceIds =
        device != null ? Collections.singletonList(device) : Arrays.asList("dev1", "dev2");
    List<String> measurements = Collections.singletonList(mea);
    ResultSet rs = r.queryByRow(deviceIds, measurements, offset, limit);
    int count = 0;
    while (rs.next()) {
      count++;
    }
    rs.close();
    return count;
  }

  /**
   * Write a tree-model TsFile with the given devices and measurement. Timestamps are 0..numRows-1,
   * values are timestamp * 10 (INT64).
   */
  private static void writeTreeFile(
      String filePath, List<String> deviceIds, String measurement, int numRows)
      throws IOException, WriteProcessException {
    File file = new File(filePath);
    MeasurementSchema schema = new MeasurementSchema(measurement, TSDataType.INT64);
    try (TsFileTreeWriter writer = new TsFileTreeWriterBuilder().file(file).build()) {
      for (String deviceId : deviceIds) {
        writer.registerTimeseries(deviceId, schema);
      }
      for (String deviceId : deviceIds) {
        for (int i = 0; i < numRows; i++) {
          TSRecord record = new TSRecord(deviceId, i);
          record.addPoint(measurement, (long) i * 10);
          writer.write(record);
        }
      }
    }
  }
}

File: cpp/test/reader/tree_view/tsfile_tree_reader_row_query_test.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include <gtest/gtest.h>

#include <climits>
#include <random>

#include "common/config/config.h"
#include "common/record.h"
#include "common/schema.h"
#include "file/write_file.h"
#include "reader/result_set.h"
#include "reader/tsfile_tree_reader.h"
#include "writer/tsfile_tree_writer.h"

using namespace storage;
using namespace common;

// ─────────────────────────────────────────────────────────────────────────────
// Fixture
// ─────────────────────────────────────────────────────────────────────────────

class TsFileTreeReaderRowQueryTest : public ::testing::Test {
   protected:
    static std::string generate_random_string(int length) {
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_int_distribution<> dis(0, 61);
        const std::string chars =
            "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
        std::string s;
        for (int i = 0; i < length; ++i) s += chars[dis(gen)];
        return s;
    }

    void SetUp() override {
        libtsfile_init();
        file_name_ = "tsfile_tree_row_query_test_" +
                     generate_random_string(10) + ".tsfile";
        remove(file_name_.c_str());
    }

    void TearDown() override {
        remove(file_name_.c_str());
        libtsfile_destroy();
    }

    // Write `num_rows` records for each device in `device_ids`,
    // using measurement `mea` (INT64). Timestamps are 0..num_rows-1,
    // values are timestamp * 10.
    void write_data(const std::vector<std::string>& device_ids,
                    const std::string& mea, int num_rows) {
        WriteFile wf;
        int flags = O_WRONLY | O_CREAT | O_TRUNC;
#ifdef _WIN32
        flags |= O_BINARY;
#endif
        wf.create(file_name_, flags, 0666);
        TsFileTreeWriter writer(&wf);
        for (auto dev :
             device_ids) {  // copy: register_timeseries needs non-const ref
            auto* schema = new MeasurementSchema(mea, INT64);
            writer.register_timeseries(dev, schema);
            delete schema;
            for (int i = 0; i < num_rows; i++) {
                TsRecord record(dev, static_cast<int64_t>(i));
                record.add_point(mea, static_cast<int64_t>(i * 10));
                writer.write(record);
            }
        }
        writer.flush();
        writer.close();
    }

    // Count the number of rows returned by queryByRow.
    int count_rows(const std::string& file,
                   const std::vector<std::string>& devs, const std::string& mea,
                   int offset, int limit) {
        TsFileTreeReader reader;
        EXPECT_EQ(reader.open(file), E_OK);
        ResultSet* rs = nullptr;
        EXPECT_EQ(reader.queryByRow(devs, {mea}, offset, limit, rs), E_OK);
        EXPECT_NE(rs, nullptr);
        int count = 0;
        bool has_next = false;
        while (IS_SUCC(rs->next(has_next)) && has_next) count++;
        reader.destroy_query_data_set(rs);
        reader.close();
        return count;
    }

    std::string file_name_;
    enum { TOTAL_ROWS = 50 };
    const std::string DEV = "device";
    const std::string MEA = "s1";
};

// ─────────────────────────────────────────────────────────────────────────────
// Tests
// ─────────────────────────────────────────────────────────────────────────────

// ① limit = 0: empty result
TEST_F(TsFileTreeReaderRowQueryTest, LimitZeroReturnsEmpty) {
    write_data({DEV}, MEA, TOTAL_ROWS);
    ASSERT_EQ(count_rows(file_name_, {DEV}, MEA, 0, 0), 0);
}

// ② limit < total: only `limit` rows returned
TEST_F(TsFileTreeReaderRowQueryTest, LimitLessThanTotal) {
    write_data({DEV}, MEA, TOTAL_ROWS);
    ASSERT_EQ(count_rows(file_name_, {DEV}, MEA, 0, 20), 20);
}

// ③ limit > total: all rows returned
TEST_F(TsFileTreeReaderRowQueryTest, LimitExceedsTotal) {
    write_data({DEV}, MEA, TOTAL_ROWS);
    ASSERT_EQ(count_rows(file_name_, {DEV}, MEA, 0, 1000), TOTAL_ROWS);
}

// ④ limit < 0: unlimited, returns all rows
TEST_F(TsFileTreeReaderRowQueryTest, NegativeLimitMeansUnlimited) {
    write_data({DEV}, MEA, TOTAL_ROWS);
    ASSERT_EQ(count_rows(file_name_, {DEV}, MEA, 0, -1), TOTAL_ROWS);
}

// ⑤ offset + limit in the middle of the data
TEST_F(TsFileTreeReaderRowQueryTest, OffsetPlusLimit) {
    write_data({DEV}, MEA, TOTAL_ROWS);
    // offset=10, limit=15 → should return exactly 15 rows
    ASSERT_EQ(count_rows(file_name_, {DEV}, MEA, 10, 15), 15);
}

// ⑥ offset >= total: empty result
TEST_F(TsFileTreeReaderRowQueryTest, OffsetBeyondTotal) {
    write_data({DEV}, MEA, TOTAL_ROWS);
    ASSERT_EQ(count_rows(file_name_, {DEV}, MEA, 1000, 10), 0);
}

// ⑦ offset + limit > total: return remaining rows from offset
TEST_F(TsFileTreeReaderRowQueryTest, OffsetPlusLimitExceedsTotal) {
    write_data({DEV}, MEA, TOTAL_ROWS);
    // offset=40, limit=20 → only 10 rows remain
    ASSERT_EQ(count_rows(file_name_, {DEV}, MEA, 40, 20), 10);
}

// ⑧ Data correctness: verify timestamps start from `offset`
TEST_F(TsFileTreeReaderRowQueryTest, OffsetDataCorrectness) {
    write_data({DEV}, MEA, TOTAL_ROWS);
    TsFileTreeReader reader;
    ASSERT_EQ(reader.open(file_name_), E_OK);
    ResultSet* rs = nullptr;
    ASSERT_EQ(reader.queryByRow({DEV}, {MEA}, 5, 10, rs), E_OK);

    int count = 0;
    bool has_next = false;
    while (IS_SUCC(rs->next(has_next)) && has_next) {
        int64_t ts = rs->get_row_record()->get_timestamp();
        EXPECT_EQ(ts, static_cast<int64_t>(5 + count));
        int64_t val = rs->get_value<int64_t>(2);
        EXPECT_EQ(val, (5 + count) * 10);
        count++;
    }
    EXPECT_EQ(count, 10);
    reader.destroy_query_data_set(rs);
    reader.close();
}

// ⑨ Cross-chunk: small page size to force multiple chunks, verify correctness
TEST_F(TsFileTreeReaderRowQueryTest, CorrectnessAcrossChunks) {
    int prev = g_config_value_.page_writer_max_point_num_;
    g_config_value_.page_writer_max_point_num_ = 5;  // tiny pages

    write_data({DEV}, MEA, 30);

    TsFileTreeReader reader;
    ASSERT_EQ(reader.open(file_name_), E_OK);
    ResultSet* rs = nullptr;
    ASSERT_EQ(reader.queryByRow({DEV}, {MEA}, 5, 10, rs), E_OK);

    int count = 0;
    bool has_next = false;
    while (IS_SUCC(rs->next(has_next)) && has_next) {
        int64_t ts = rs->get_row_record()->get_timestamp();
        EXPECT_EQ(ts, static_cast<int64_t>(5 + count));
        count++;
    }
    EXPECT_EQ(count, 10);
    reader.destroy_query_data_set(rs);
    reader.close();

    g_config_value_.page_writer_max_point_num_ = prev;
}

// ⑩ Multiple devices: verify total row count with offset/limit
TEST_F(TsFileTreeReaderRowQueryTest, MultipleDevicesOffsetLimit) {
    // device_1 and device_2 each have 20 rows with timestamps 0..19.
    // QDSWithoutTimeGenerator merges them by time: at each timestamp both
    // devices' values appear in one RowRecord, so total distinct timestamps
    // = 20.
    write_data({"device_1", "device_2"}, MEA, 20);

    ASSERT_EQ(count_rows(file_name_, {"device_1", "device_2"}, MEA, 5, 10), 10);
}

// ⑪ queryByRow result set supports metadata inspection
TEST_F(TsFileTreeReaderRowQueryTest, MetadataAccessible) {
    write_data({DEV}, MEA, 10);

    TsFileTreeReader reader;
    ASSERT_EQ(reader.open(file_name_), E_OK);
    ResultSet* rs = nullptr;
    ASSERT_EQ(reader.queryByRow({DEV}, {MEA}, 0, 5, rs), E_OK);

    auto meta = rs->get_metadata();
    ASSERT_NE(meta, nullptr);
    // column 1 = "time", column 2 = measurement
    EXPECT_EQ(meta->get_column_name(1), "time");
    EXPECT_EQ(meta->get_column_count(), 2u);

    reader.destroy_query_data_set(rs);
    reader.close();
}

// ⑫ Paging consistency: two pages together equal the full result
TEST_F(TsFileTreeReaderRowQueryTest, PaginationConsistency) {
    write_data({DEV}, MEA, 40);

    int page1 = count_rows(file_name_, {DEV}, MEA, 0, 20);
    int page2 = count_rows(file_name_, {DEV}, MEA, 20, 20);
    ASSERT_EQ(page1 + page2, 40);
}
